### ISSUE TITLE
Add dropout for neural networks

### ISSUE THREAD

**Issue by alexrobbins** [encog/encog-java-core#130]
Dropout is the concept of training a neural network whose hidden layer neurons have only a .5 probability of being present in any training run. The end result is that you are effectively training an ensemble of neural networks that have massive weight sharing. This speeds training greatly (relative to training the entire ensemble separately), and eliminates some of the overfitting often associated with neural nets. For instance, neural nets with dropout never need to have their training stopped early to avoid overfitting.

Discussed in this paper: http://arxiv.org/abs/1207.0580
And in this Google tech talk: http://www.youtube.com/watch?v=DleXA5ADG78

I'd be willing to contribute code if this is something you'd be willing to include. Thanks!

**Comment by nitbix** [comment]
Hi Jeff,

Just to let you know that I've merged the latest master HEAD into
https://github.com/nitbix/encog-java-core master HEAD, so finishing the
pull request should be a lot less work for you.

Best,

Alan

On 27 February 2014 at 09:26, Alan Mosca <nitbix@nitbix.com> wrote:

> Hi Jeff,
>
> I completely understand. If you need any help, or want to point out some
> of the breakages I can try to fix them.
>
> Thanks!
>
>
> On 27 February 2014 04:14, Jeff Heaton <notifications@github.com> wrote:
>
>> By the way, I do hope to get this merged into the next Encog version. I
>> just have not had the time. There are quite a few changes, and some
>> breaking to existing programs based on Encog... So adding is non-trivial.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/encog/encog-java-core/issues/130#issuecomment-36208714>
>> .
>>
>
>
>
> --
>   Alan
>



-- 
  Alan

**Comment by jeffheaton** [comment]
Thank you very much!  I will take a look.  This will be my first priority
on the next Encog version.

Also, my paper on Encog was accepted by the JMLR.  I cited your masters
thesis for the dropout/regularization that you added to the Encog project.
Is this how you wanted it?  I can still change it, I will be in the final
publication queue for several months, I am sure.  Here is the bibtex I used.

@mastersthesis{Mosca:mscdissertation,
  author = {Mosca, A.},
  title = {Extending Encog: A study on classifier ensemble techniques},
  school = {Birkbeck, University of London},
  year = {2012}
}

**Comment by nitbix** [comment]
Hi Jeff,

Thanks for the citation, that's perfect. Let me know if there's anything
else I can do to help with getting everything merged.

Thanks,

Alan

**Comment by thorinii** [comment]
Hi all,
Did this get completed?

**Comment by nitbix** [comment]
Yes, in master the constructor for BasicLayer now takes an additional
dropoutRate, which gets used by the training algorithms.

On 19 August 2015 at 09:24, Lachlan Phillips <notifications@github.com>
wrote:

> Hi all,
> Did this get completed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/encog/encog-java-core/issues/130#issuecomment-132489889>
> .
>



-- 
  Alan

**Comment by jeffheaton** [comment]
This was merged with Encog previously.
